templates/ovs/bin/get_net_ip

```diff
@@ -1,63 +0,0 @@
-#!/bin/env python3
-#
-# Copyright 2023 Red Hat Inc.
-#
-# Licensed under the Apache License, Version 2.0 (the "License"); you may
-# not use this file except in compliance with the License. You may obtain
-# a copy of the License at
-#
-#      http://www.apache.org/licenses/LICENSE-2.0
-#
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
-# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
-# License for the specific language governing permissions and limitations
-# under the License.
-
-import os
-import json
-import sys
-
-# Uses network status from pod:
-#   metadata.annotations['k8s.v1.cni.cncf.io/networks-status']
-#
-# Example network status json string:
-#
-# [
-#   { "name": "openshift-sdn",
-#     "interface": "eth0",
-#     "ips": [ "10.131.0.76" ],
-#     "default": true, "dns": {}
-#   },
-#   { "name": "openstack/baremetal-net",
-#     "interface": "net1",
-#     "ips": [ "192.168.24.1" ],
-#     "mac": "5e:ff:15:d9:66:84",
-#     "dns": {}
-#   }
-# ]
-#
-
-def get_ip_from_network_status():
-    namespace = os.environ.get('PodNamespace')
-    net_status = json.loads(os.environ.get('PodNetworksStatus'))
```

Comment:
If this value is not there on slow systems immediately ... but appear after some time, we may want to use tenacity.retry here?

Reply:
I'll give this a try

Reply:
@hjensas thinking of it again and also tried it, this does not work. A pods env var won't get updated while its running. If the value  changes, the pod needs to be restarted.

```
+ OvnBridge=br-int
+ OvnRemote=tcp:172.17.0.31:6642
+ OvnEncapType=geneve
+ '[' -n internalapi ']'
++ /usr/local/bin/container-scripts/get_net_ip internalapi
2023-03-09 07:38:22 INFO     0 attempt to load json from PodNetworksStatus: 
2023-03-09 07:38:22 ERROR    Error loading json from PodNetworksStatus in 0 try:
2023-03-09 07:38:27 INFO     1 attempt to load json from PodNetworksStatus: 
2023-03-09 07:38:27 ERROR    Error loading json from PodNetworksStatus in 1 try:
2023-03-09 07:38:32 INFO     2 attempt to load json from PodNetworksStatus: 
2023-03-09 07:38:32 ERROR    Error loading json from PodNetworksStatus in 2 try:
2023-03-09 07:38:37 INFO     3 attempt to load json from PodNetworksStatus: 
2023-03-09 07:38:37 ERROR    Error loading json from PodNetworksStatus in 3 try:

...

2023-03-09 07:39:02 INFO     8 attempt to load json from PodNetworksStatus: 
2023-03-09 07:39:02 ERROR    Error loading json from PodNetworksStatus in 8 try:
2023-03-09 07:39:07 INFO     9 attempt to load json from PodNetworksStatus: 
Traceback (most recent call last):
  File "/usr/local/bin/container-scripts/get_net_ip", line 62, in get_ip_from_network_status
    net_status = json.loads(pod_network_status)
  File "/usr/lib64/python3.9/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python3.9/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python3.9/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/container-scripts/get_net_ip", line 94, in <module>
    get_ip_from_network_status()
  File "/usr/local/bin/container-scripts/get_net_ip", line 65, in get_ip_from_network_status
    raise Exception(
Exception: Could not load json data from PodNetworksStatus environment var: pod_network_status

```

Right now I these options:
* pass in the interface index and enhance the script for ipv6 use case, like excluding the link local address
  `OvnEncapIP=$(ip a s dev ${OvnEncapNIC} | grep -v "scope link" | grep "inet" | awk '{print $2}' | cut -d "/" -f 1)`
* make the init script to fail if a valid IP could not be identified and get the pod restarted and "hope" the env var is correct set on next pod created
* use retry like I tested, but then have the operator to exec into the pod and update the env var if required, but this is mehhh ...
* allow the container to query the OCP api to get the annotation via an api call. I also don't think this is what we want.

Reply:
> Right, I was living on a hope that the env would update by magic of some volume mount or something.
> 
> This seems like a race bug that kubernetes should handle better? I do wonder, can we check that the annotation is present in the DaemonSet spec - and re-reconcile until it is? I.e we don't create the init pods unless it is there? https://github.com/openstack-k8s-operators/ovs-operator/blob/main/pkg/ovs/daemonset.go#L99-L103 https://github.com/openstack-k8s-operators/ovs-operator/blob/main/controllers/ovs_controller.go#L334
> 
> Edit: Or even earlier in lib-common networkattachment where `CreateNetworksAnnotation` is implemented? Edit_again: I guess we cannot check, because the data is not created before the demonset creates the pods.

right, thats not possible. the Daemoset creates the pod and we have not control to wait for the information.

> 
> I wonder if we use a downwardAPI volume instead ... there is a note in docs: `Note: A container using the downward API as a [subPath](https://kubernetes.io/docs/concepts/storage/volumes/#using-subpath) volume mount does not receive updates when field values change.` i.e if we mount a /downward_api volume then the information on that volume should update as values change?
> 
> (I guess I'm worried about using index because of historical enumerations instability :) )

@hjensas Its possible to request an interface name when. I'll submit a PR I'll submit a PR for CreateNetworksAnnotation to when create the annotation to request the if name to be the NAD. only thing to consider is that the max IF name length is 16 char if I am correct. how does that sound?

Reply:
> I'd suggest we follow up on a the and for now use the interface with name pattern for the short term. this is the PR to change lib-common for this [openstack-k8s-operators/lib-common#214](https://github.com/openstack-k8s-operators/lib-common/pull/214) .

Agreed.